pkg/operation/botanist/controlplane/kubecontrollermanager/waiter.go

```diff
@@ -0,0 +1,89 @@
+	"github.com/gardener/gardener/pkg/operation/common"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (k *kubeControllerManager) Wait(_ context.Context) error        { return nil }
+func (k *kubeControllerManager) WaitCleanup(_ context.Context) error { return nil }
+
+func (k *kubeControllerManager) WaitForControllerToBeActive(ctx context.Context) error {
+	const (
+		pollInterval = 5 * time.Second
+		pollTimeout  = 90 * time.Second
+	)
+
+	// Check whether the kube-controller-manager deployment exists
+	if err := k.seedClient.Get(ctx, kutil.Key(k.namespace, v1beta1constants.DeploymentNameKubeControllerManager), &appsv1.Deployment{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Errorf("kube controller manager deployment not found: %v", err)
+		}
+		return err
+	}
+
+	return retry.UntilTimeout(ctx, pollInterval, pollTimeout, func(ctx context.Context) (done bool, err error) {
+		podList := &corev1.PodList{}
+		err = k.seedClient.List(ctx, podList,
+			client.InNamespace(k.namespace),
+			client.MatchingLabels(map[string]string{
+				v1beta1constants.LabelApp:  v1beta1constants.LabelKubernetes,
+				v1beta1constants.LabelRole: labelRole,
+			}))
+		if err != nil {
+			return retry.SevereError(fmt.Errorf("could not check whether controller %s is active: %w", v1beta1constants.DeploymentNameKubeControllerManager, err))
+		}
+
+		// Check that only one replica of the controller exists.
+		if len(podList.Items) != 1 {
+			k.log.Infof("Waiting for %s to have exactly one replica", v1beta1constants.DeploymentNameKubeControllerManager)
+			return retry.MinorError(fmt.Errorf("controller %s is not active", v1beta1constants.DeploymentNameKubeControllerManager))
+		}
+
+		// Check that the existing replica is not getting deleted.
+		if podList.Items[0].DeletionTimestamp != nil {
+			k.log.Infof("Waiting for a new replica of %s", v1beta1constants.DeploymentNameKubeControllerManager)
+			return retry.MinorError(fmt.Errorf("controller %s is not active", v1beta1constants.DeploymentNameKubeControllerManager))
+		}
+
+		// Check if the controller is active by reading its leader election record.
+		leaderElectionRecord, err := common.ReadLeaderElectionRecord(ctx, k.shootClient, resourcelock.EndpointsResourceLock, metav1.NamespaceSystem, v1beta1constants.DeploymentNameKubeControllerManager)
+		if err != nil {
+			return retry.SevereError(fmt.Errorf("could not check whether controller %s is active: %w", v1beta1constants.DeploymentNameKubeControllerManager, err))
+		}
+
+		if delta := metav1.Now().UTC().Sub(leaderElectionRecord.RenewTime.Time.UTC()); delta <= pollInterval-time.Second {
```

Comment:
Can you help me understanding this by explaining what this line does?

Reply:
I did not create this LoC but merely moved the code block.
With the help of @rfranzke : added a comment explaining it.

Reply:
Thanks that makes it clearer. I don't yet understand how the pollInterval is related to that though.

WDYT about explaining it in code instead?
```
  lastRenew:= metav1.Now().UTC().Sub(leaderElectionRecord.RenewTime.Time.UTC()
  defaultRenewalTime := 2*time.Second
  if lastRenew <= defaultRenewalTime * 2 {
    return retry.Ok()
  }
```
cc @rfranzke 



return retry.Ok()
]
```

Reply:
Looks good to me - added the code in the way suggested.

Reply:
I would delete the comment then as it is now redundant 😉